cupyx/scipy/special/_gammainc.py

```diff
@@ -0,0 +1,1202 @@
+Copyright 1984, 1987 by Stephen L. Moshier
+Direct inquiries to 30 Frost Street, Cambridge, MA 02140
+"""
+
+from cupy import _core
+
+from cupyx.scipy.special._digamma import polevl_definition
+from cupyx.scipy.special._zeta import zeta_definition
+
+_zeta_c = zeta_definition
+# comment out duplicate MACHEP definition
+_zeta_c = _zeta_c.replace(
+    "__constant__ double MACHEP", "// __constant__ double MACHEP"
+)
+
+
+# TODO: finish
+
+# See the following on error states:
+#    docs.scipy.org/doc/scipy/reference/generated/scipy.special.errstate.html
+#    scipy/special/sf_error.h
+
+# TODO: which value for MACHEP, MAXLOG from const.c to use?
+
+_misc_preamble = """
+
+// include for CUDART_NAN, CUDART_INF
+#include <cupy/math_constants.h>
+
+// defines from /scipy/special/cephes/const.c
+__constant__ double MACHEP = 1.11022302462515654042E-16;
+__constant__ double MAXLOG = 7.09782712893383996732E2;
+
+// defines from numpy/core/include/numpy/npy_math.h
+#define NPY_EULER     0.577215664901532860606512090082402431  /* Euler constant */
+#define NPY_PI        3.141592653589793238462643383279502884  /* pi */
+#define NPY_SQRT2     1.414213562373095048801688724209698079  /* sqrt(2) */
+#define NPY_SQRT1_2   0.707106781186547524400844362104849039  /* 1/sqrt(2) */
+
+// from /scipy/special/cephes/mconf.h
+#ifdef MAXITER
+#undef MAXITER
+#endif
+
+#define MAXITER 2000
+#define IGAM 1
+#define IGAMC 0
+#define SMALL 20
+#define LARGE 200
+#define SMALLRATIO 0.3
+#define LARGERATIO 4.5
+
+__constant__ double big = 4.503599627370496e15;
+__constant__ double biginv = 2.22044604925031308085e-16;
+
+"""
+
+
+# Note: p1evl -> followed the template style used for polevl in digamma.py
+p1evl_definition = """
+template<int N> static __device__ double p1evl(double x, const double coef[])
+{
+    double ans;
+    const double *p;
+    int i;
+
+    p = coef;
+    ans = x + *p++;
+
+    for (i = 0; i < N - 1; ++i){
+        ans = ans * x + *p++;
+    }
+
+    return ans;
+}
+"""
+
+lgam_definition = """
+// define lgam, using CUDA's lgamma as done for cupyx.scipy.special.gammaln
+__device__ double lgam(double x)
+{
+    if (isinf(x) && x < 0) {
+        return -1.0 / 0.0;
+    } else {
+        return lgamma(x);
+    }
+}
+"""
+
+# lgam1p from unity.c
+_unity_c_partial = (
+    _zeta_c
+    + lgam_definition
+    + polevl_definition
+    + p1evl_definition
+    + """
+// part of cephes/unity.c
+
+__constant__ double LP[] = {
+    4.5270000862445199635215E-5,
+    4.9854102823193375972212E-1,
+    6.5787325942061044846969E0,
+    2.9911919328553073277375E1,
+    6.0949667980987787057556E1,
+    5.7112963590585538103336E1,
+    2.0039553499201281259648E1,
+};
+
+__constant__ double LQ[] = {
+    /* 1.0000000000000000000000E0, */
+    1.5062909083469192043167E1,
+    8.3047565967967209469434E1,
+    2.2176239823732856465394E2,
+    3.0909872225312059774938E2,
+    2.1642788614495947685003E2,
+    6.0118660497603843919306E1,
+};
+
+__device__ double log1p(double x)
```

Comment:
Can this function be replaced by CUDA [log1p](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__DOUBLE.html#group__CUDA__MATH__DOUBLE_1g3c680d660d75780ef53075a439211626) or [log1pf](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__SINGLE.html#group__CUDA__MATH__SINGLE_1g9d53128ab5f7d6ebc4798f243481a6d7)?

Reply:
Yes, I tested that the CUDA Math `log1p` still passes the tests. I can make a small PR to drop the redundant code